### ISSUE TITLE
[css-fonts-5] advance-override details

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#5983]
We resolved to add an advance-override descriptor in https://github.com/w3c/csswg-drafts/issues/5533 but didn't define quite what it means. Here are some fun questions:

* How are we increasing/decreasing the advance width? Are we increasing the font size? Stretching in the X axis? Adding space around each glyph? Something else?
* If you're adding space around glyphs, is it applied at the start/end of the line also? On Arabic and Mongolian as well? etc.

**Comment by xiaochengh** [comment]
> - How are we increasing/decreasing the advance width? Are we increasing the font size? Stretching in the X axis? Adding space around each glyph? Something else?

In Chrome's implementation, there's no font size changes or stretches. It shares quite a lot of code with `letter-spacing`, so it ends up adding space after each glyph.

IMO this seems better than adding spacing around each glyph. Otherwise, the first character may look misaligned with the start of the line.

Anyway, all these choices have no difference for the purpose of reducing layout shift...

> - If you're adding space around glyphs, is it applied at the start/end of the line also?

This is one of the topics that apply to `letter-spacing` but not to `advance-override`, since we are modifying the advances of the glyphs instead of adding spacing in between. I don't see a reason why we should leave the start/end of line unmodified.

> - On Arabic and Mongolian as well? etc.

Maybe we should recommend NO for complex scripts and fonts. Reasons:
- The legibility really depends on whether the text shaper supports changing glyph widths, which is case by case different.
- In [some](https://jsfiddle.net/wpnjav2k/1/) [cases](https://jsfiddle.net/mgx5t72s/), different fonts can shape the same text with different numbers of glyphs. Then `advance-override` won't work as good as, e.g., Latin.
  - Note: I can't read these languages, so I don't know if this is a common case, or due to the generated gibberish text

**Comment by chrishtr** [comment]
+1 to skipping scripts with cursive representations. I think this is in accord with the F2F discussion also, and is what `letter-spacing` does (*)


(*) Spec text: "if the UA cannot expand text from a cursive script without breaking its cursive connections, it must not apply spacing between any pair of that script’s typographic letter units at all" https://drafts.csswg.org/css-text-3/#cursive-tracking

**Comment by r12a** [comment]
> Note: I can't read these languages, so I don't know if this is a common case, or due to the generated gibberish text

It is indeed a common case for complex scripts to use different letter shapes, font styles, and ligating or conjoining approaches from font to font.  It is also common that the text is formed from clusters of glyphs, with complex relative offsets and rule-base variants, rather than a simple succession of isolated letter glyphs.  These things are so common, in fact, that they are what lead to them being called 'complex scripts'.

But even where there is little to no interaction between glyphs, other than placement of diacritics, such as for Thai, differences between individual letter shapes [can vary](https://r12a.github.io/scripts/thai/#writing_styles). Going from one font to another, some letters are thinner, others wider.  

Actually i'd have thought that would also be problematic for Latin text unless you choose quite similar fonts.  Would it be fair to say that advance-override is really most suitable for monospaced, non-complex fonts such as CJK?

**Comment by xiaochengh** [comment]
> But even where there is little to no interaction between glyphs, other than placement of diacritics, such as for Thai, differences between individual letter shapes can vary. Going from one font to another, some letters are thinner, others wider.

Yeah, for this reason `advance-override` cannot accurate match the text width between fonts.

We never intend to do that accurately, though. Our goal is just to reduce layout shift. Given that web developers are already using `letter-spacing` to reduce layout shift, and our proposed version works better for the same purpose, I think it should be good enough.

**Comment by svgeesus** [comment]
It sounds like the description for this should point out the limitations, and thus avoid over-promising

**Comment by xiaochengh** [comment]
Here are some topics and my proposals for the next meeting. Please add more if needed.

1. Are characters stretched/resized/etc when `advance-override` is applied?

Proposal: no. It has no effect for the purpose of reducing layout shift, but adds a lot of complexity to spec and implementation. If using `advance-override` severely affects legibility, then it's better to just use a more similar fallback font.

2. How is spacing added to each character?

Proposal: spacing is added after each character, for the reasons I gave in https://github.com/w3c/csswg-drafts/issues/5983#issuecomment-777934132 for reasons.

3. What's the default value of the second parameter, and what's the exact syntax?

Proposal: Use two descriptors `advance-width-override` and `advance-height-override` separately. This seems the cleanest among all the ideas I came up with.

Other ideas:
- `normal | [ <percentage> <percentage> ]` as [currently specified](https://drafts.csswg.org/css-fonts-5/#descdef-font-face-advance-override). In other words, second parameter is the same as the first if unspecified.
- `[normal | <percentage>] [normal | percentage]?`. In other words, it has no override to upright vertical text if the second parameter is not explicitly given.

4. Add spec text to avoid over-promising. This includes:
   - It can't fully eliminate layout shift, unless the font is monospaced
   - It may not work on certain cursive/complex scripts, [similar to `letter-spacing`](https://drafts.csswg.org/css-text-3/#cursive-tracking)
   - (Add more here...)

**Comment by r12a** [comment]
> It may not work on certain cursive/complex scripts, similar to letter-spacing

Just a nit:

It's similar to letter-spacing for _cursive_ scripts (although there's still a remote possibility that someone might come up with a use case that makes sense, esp. for non-Arabic scripts). 

I don't think however that it's similar to letter-spacing wrt complex scripts in general, because letter-spacing is expected to handle those (albeit that some additional functionality needs to be specified/implemented*).

hth

* If you're curious about the details, see the latter part of the article we currently have out for review https://w3c.github.io/i18n-drafts/questions/qa-indic-graphemes.en.html.

**Comment by jfkthame** [comment]
> Here are some topics and my proposals for the next meeting. Please add more if needed.
> 
>     1. Are characters stretched/resized/etc when `advance-override` is applied?
> 
> Proposal: no. It has no effect for the purpose of reducing layout shift, but adds a lot of complexity to spec and implementation. If using `advance-override` severely affects legibility, then it's better to just use a more similar fallback font.

Agreed, no. `advance-override` doesn't touch glyph images, only spacing.

(I think there should be a separate `font-size-override` that can apply a scale factor to the used font size; I suspect this would often be the better tool to use.)

>     1. How is spacing added to each character?
> 
> Proposal: spacing is added after each character, for the reasons I gave in [#5983 (comment)](https://github.com/w3c/csswg-drafts/issues/5983#issuecomment-777934132) for reasons.

Disagree; I think spacing should be added equally on both sides of each glyph. Adding it asymmetrically (as some current letter-spacing implementations do) is often visually inferior IMO, particularly at run boundaries.

In general this feature should not be used to insert (or remove) more than a small percentage of the glyph width -- if you need more than that, you're using a bad choice of fallback font, or should be using `font-size-override` (see above) to completely re-scale it rather than just munging the spacing.

>     1. What's the default value of the second parameter, and what's the exact syntax?
> 
> Proposal: Use two descriptors `advance-width-override` and `advance-height-override` separately. This seems the cleanest among all the ideas I came up with.
> 
> Other ideas:
> 
>     * `normal | [ <percentage> <percentage> ]` as [currently specified](https://drafts.csswg.org/css-fonts-5/#descdef-font-face-advance-override). In other words, second parameter is the same as the first if unspecified.
> 
>     * `[normal | <percentage>] [normal | percentage]?`. In other words, it has no override to upright vertical text if the second parameter is not explicitly given.

I don't have a strong preference among these options. I think I'd lean towards either the first (separate width/height descriptors) or the third (second param defaults to `normal`, so no height override unless explicitly given), as most authors won't be thinking about vertical layout at all unless it's their primary focus, and there's little reason to expect that the *same* amount of override will make sense for both axes.

>     1. Add spec text to avoid over-promising. This includes:
>        
>        * It can't fully eliminate layout shift, unless the font is monospaced
>        * It may not work on certain cursive/complex scripts, [similar to `letter-spacing`](https://drafts.csswg.org/css-text-3/#cursive-tracking)

+1 to mentioning that it may not work well with cursive/complex scripts.

However, I think we should *not* have spec text (like for `letter-spacing`) saying this doesn't apply to joined scripts, or anything "sophisticated" like that. This is equivalent to loading the font resource into a font editor and adding/removing sidebearings on each glyph to change its advance by a given percentage. Whether/how those glyphs interact with each other is not our concern here. This is an operation at the level of glyphs, not Unicode scripts/characters.

By all means include a warning for authors about applying this to glyphs that are supposed to visually join, but if an author nevertheless chooses to do so, they can. (I have seen old printed examples from lead-type days of `letter-spacing`-like effects in what would usually be joined scripts. Whether it's a good idea is not for us to say.)

If an author wants to (for example) use Arial as a fallback, with `advance-override` applied to its Latin letters but not to Arabic, they can do that by using `unicode-range` to spread it across a couple of `@font-face` rules, applying different descriptors to the different blocks of characters.

**Comment by litherum** [comment]
> +1 to skipping scripts with cursive representations.

I thought the reason we decided it would add a constant to each advance was because that looked best for cursive scripts? Maybe I'm misremembering?

I also don't understand why the behavior here is different than `letter-spacing`. Philosophically, what about these two things means they should behave differently?

Corollary: I think my counter-proposal to the above discussion would be: "intentionally don't innovate here, and make it behave exactly like `letter-spacing`."

**Comment by jfkthame** [comment]
[`letter-spacing` as currently defined in CSS](https://drafts.csswg.org/css-text-3/#letter-spacing-property) (whether it's a good definition may be a topic for a separate discussion) isn't something that can readily be specified/implemented as an attribute of a particular font resource, because of how it behaves at element boundaries. Given `<span style="letter-spacing:1em">foo</span>` you can't just adjust the metrics of the `f`, `o`, `o` and be done; what happens at the edges is supposed to depend on the style of the *parent* element.

That would even come into play if you have two *immediately adjacent* `<span>`s with the same `letter-spacing` property: what happens at the boundary between them isn't the result of their `letter-spacing` but of their common parent's:

> Letter spacing between two typographic character units effectively “belongs” to the innermost element that contains the two typographic character units: the total letter spacing between two adjacent typographic character units (after bidi reordering) is specified by and rendered within the innermost element that contains the boundary between the two typographic character units.

I think `advance-override` should explicitly *not* attempt to have this kind of behavior; to serve the intended purpose of making the metrics of a fallback font more similar to another "primary" font, it must simply apply, unconditionally, to the individual glyphs rendered by that font.

Now, if you want to support *changing* the definition of `letter-spacing` (which AFAIK none of the major browsers currently implements as spec'd), I'd be happy to entertain that idea. But as it stands, I am definitely opposed to saying that `advance-override` should behave similarly to how `letter-spacing` is spec'd. I believe it belongs to a different level of the rendering model.

**Comment by litherum** [comment]
> you can't just adjust the metrics

Right, I'm not proposing just adjusting the metrics. I'm proposing an implementation which is more involved than simply overwriting some values from the file.

**Comment by fantasai** [comment]
A couple points:

> I think there should be a separate font-size-override that can apply a scale factor to the used font size; I suspect this would often be the better tool to use.

Agreed, and I'd go further: I think shipping `advance-override` without shipping `font-size-override` would be harmful. Because advance-override breaks a lot of writing systems and can create weird rhythms in the text that are likely to impact readability/accessibility, and font-size-override wouldn't have those problems. In any situation were font-size-override would be adequate, it should be used instead of advance-override.

> I think I'd lean towards either the first (separate width/height descriptors) or the third (second param defaults to normal, so no height override unless explicitly given), as most authors won't be thinking about vertical layout at all unless it's their primary focus, and there's little reason to expect that the same amount of override will make sense for both axes.

Agreed. I'd probably go with the third option, seems the friendliest syntactically.

>>>    How is spacing added to each character?
>>  Proposal: spacing is added after each character, for the reasons I gave in #5983 (comment) for reasons.
> Disagree; I think spacing should be added equally on both sides of each glyph. [[ rationale ]](https://github.com/w3c/csswg-drafts/issues/5983#issuecomment-788776441)

Agree with jfkthame's rationale. Yes, adding it just to the right side works best for left-aligned paragraphs. But not all affected content is left-aligned.

> I think my counter-proposal to the above discussion would be: "intentionally don't innovate here, and make it behave exactly like letter-spacing

Note that letter-spacing adds a constant per typographic character unit. Advance-override is defined to scale the advance of each glyph proportionally. Also it's being defined as a font metric override, not a text layout feature...

**Comment by litherum** [comment]
After discussing with the rest of the team, I retract my proposal. Sorry for the noise.

**Comment by litherum** [comment]
One thing that is important to note here - these overridden advances will have to be applied after shaping, not before. If the input to shaping doesn't match what's in the font file, bad things happen (see https://bugs.webkit.org/show_bug.cgi?id=214769).

**Comment by xiaochengh** [comment]
I don't have a strong opinion on where to add spacing: before, after or around each glyph. All work the same for the purpose of reducing layout shift, so I'm fine to accept any resolution that gives the best typography.

> Agreed, and I'd go further: I think shipping `advance-override` without shipping `font-size-override` would be harmful.

We actually have prototyped `font-size-override` in Chrome, but dropped it from the proposal because it seems tricky to cleanly specify:
- Should `ascent-override` etc resolve their percentage value against the `font-size` property value, or the overridden font size? Either way seems not very natural.
- Should the resolution of length units `ex`, `ch` and etc be affected?

I'm happy to see if `font-size-override` can be specified nicely, since it gives an alternative to reduce layout shift, but without breaking shaping.

**Comment by fantasai** [comment]
@xiaochengh No, no units should be affected by font-size override. It's a scaling factor applied to the glyphs and metrics, but not to CSS. as for how it interacts with other overrides, probably it should apply first.

**Comment by fantasai** [comment]
@xiaochengh Actually, thinking about it more, it shouldn't affect 'em' but probably should affect things that measure off the glyph like ex and ic. (There's a lot of use cases for per-font scaling adjustments that aren't about fallback.)

**Comment by fantasai** [comment]
For this issue based on the discussion above I propose that we
- Define `advance-override` to take two values, second one for the vertical axis defaulting to `normal`.
- Apply extra space on both sides of the glyph so that inline runs of affected text and shrink-to-fit&ndash;sized boxes aren't lopsided and so that left and right alignment are symmetrically affected.
- Clarify that the adjustment is applied unconditionally per *glyph* after shaping. This means it breaks cursive and complex scripts and has weird effects wherever there are ligatures, and the spec should warn about this.

I'm not really happy about that last one, though.

@xiaochengh @jfkthame Filed https://github.com/w3c/csswg-drafts/issues/6075 on glyph scaling, with answers to most of the questions I think you have. Lmk what you think.

**Comment by xiaochengh** [comment]
> Clarify that the adjustment is applied unconditionally per glyph after shaping. This means it breaks cursive and complex scripts and has weird effects wherever there are ligatures, and the spec should warn about this.

@fantasai Could you clarify how this works? Especially when we have a grapheme cluster (e.g., `A&#x0300;` or À), how do we adjust the individual glyphs there (like the diacritic)?

**Comment by jfkthame** [comment]
In the case of `A&#x0300;`, if the font doesn't form a precomposed glyph for À, the diacritic will have zero width and therefore "overriding" its width with a percentage will not have any effect. If OpenType mark attachment has been used to position it, it should maintain its position relative to the base glyph to which it is attached.

If the font didn't actually attach the glyph, but just used a zero-width glyph with negative left sidebearing, so that it "overstrikes" the preceding letter, then adjusting the advance of the base letter could have the effect of shifting the diacritic relative to its base. But I don't think that's a common situation.

(Contrary to what I understood @litherum to be saying on the recent call, my current thinking is that it should be possible to apply `advance-override` *during* (not after) the text shaping process, and this will automatically yield proper behavior for things like mark attachment within a cluster. But I have not attempted to implement this, and maybe it will break in ways I'm not currently anticipating.)

**Comment by xiaochengh** [comment]
Now I see how per glyph adjustment is supposed to work. It should be straightforward if spacing is only added behind each glyph. But, if we add spacing to both both sides, then the horizontal offset of the glyphs inside a grapheme cluster should also be adjusted. It should still work, though.

@fantasai's last [proposal](https://github.com/w3c/csswg-drafts/issues/5983#issuecomment-790241213) looks acceptable to me, and I don't have further opinions.

**Comment by yisibl** [comment]
Can we consider emoji type fonts?

**Comment by xiaochengh** [comment]
I'm [prototyping](https://chromium-review.googlesource.com/c/chromium/src/+/2747538) @fantasai's [proposal](https://github.com/w3c/csswg-drafts/issues/5983#issuecomment-790241213), which is very implementable. And I found one detail that I'd like to discuss regarding whether we break grapheme clusters.

Consider Bengali text `আগুনের শিখা নিভে গিয়েছিল` with `advance-override: 200%`. A strict implementation of the proposal gives the following, where many grapheme clusters (syllables) are broken:

![7C45F5DE-C4ED-4EDB-8CC6-24AFA7FF1973_4_5005_c](https://user-images.githubusercontent.com/19384155/111215978-0c6aa580-8591-11eb-8b9f-45095dd9418f.jpeg)

This is because a grapheme cluster (e.g., `নে`) may be composed of multiple glyphs, each of which has a positive advance. When we add extra spacing to both sides of the glyphs, they got separated.

(Btw the situation seems common in many Indic and related scripts)

---

One thing I like about the proposal is that, it affects the text width (and hence CLS) in a very predictable way: **the result text width is scaled by exactly the given percentage**. So I'd like to do some small tweaks to make the text layout look better, while preserving this property. The result on the previous Bengali text looks like:

![C9C2981F-EACD-42FF-BDF2-2A2F1693E2CE_4_5005_c](https://user-images.githubusercontent.com/19384155/111216209-5d7a9980-8591-11eb-80ff-28a1b771ff5d.jpeg)

(Note that the total text width is exactly the same as before)

The tweak is basically adjusting the total advance of each *grapheme cluster* as a whole, instead of glyphs:
1. For each grapheme cluster, let `advance` be the total advance of its glyphs
2. An extra spacing of `advance * (advance_override - 1) / 2` is applied before the first glyph and after the last glyph of the grapheme cluster, respectively

Similar as before, the adjustment is unconditionally applied to each grapheme cluster after shaping.

This also handles simpler cases like diacritics well, since diacritic glyphs have zero advance:

<img width="483" alt="Screen Shot 2021-03-15 at 2 31 07 PM" src="https://user-images.githubusercontent.com/19384155/111224036-2a3d0800-859b-11eb-99fe-5f43469c497a.png">

Unfortunately, cursive joining of e.g. Arabic may still be broken because each letter is a single grapheme cluster (at least when using Noto Naskh Arabic). This seems unfixable, because after shaping, there's no way to know whether adding spacing between two grapheme clusters breaks cursive joining or not. The situation is the same as when applying `letter-spacing` to Arabic.

<img width="435" alt="Screen Shot 2021-03-15 at 2 32 32 PM" src="https://user-images.githubusercontent.com/19384155/111224185-4e004e00-859b-11eb-9bd4-58817046c012.png">

**Comment by xiaochengh** [comment]
> Can we consider emoji type fonts?

@yisibl I think all proposals here naturally handle emoji fonts well, because emoji fonts are monospaced

**Comment by kojiishi** [comment]
fyi @r12a please see https://github.com/w3c/csswg-drafts/issues/5983#issuecomment-799770586 from @xiaochengh above.

**Comment by jfkthame** [comment]
I agree with [@xiaochengh's comment](https://github.com/w3c/csswg-drafts/issues/5983#issuecomment-799770586) about applying to grapheme clusters rather than individual characters; this will generally be less broken.

However, I also agree with [@fantasai's point](https://github.com/w3c/csswg-drafts/issues/5983#issuecomment-789898142) that `advance-override` is fundamentally broken for many writing systems. It has the potential to be a harmful feature that hurts typography and even accessibility on the web.

I believe that the use cases that prompted the invention of `advance-override` would in every case be better served by a [scaling descriptor](https://github.com/w3c/csswg-drafts/issues/6075#issuecomment-799915767), and the better way forward is for browsers to implement this.

@xiaochengh has already experimented with scaling the font size instead of overriding glyph advances, and reported that it [does seem to work better](https://github.com/w3c/csswg-drafts/issues/5533#issuecomment-714166725). I have also [prototyped such a feature in Gecko](https://bugzilla.mozilla.org/show_bug.cgi?id=1698495) and it seems to work as expected there.

In the light of this, I propose that we should *reverse* the resolution in #5533 to add the `advance-override` descriptor, and instead focus on spec'ing and shipping the [scaling descriptor](https://github.com/w3c/csswg-drafts/issues/6075#issuecomment-799915767), as being fundamentally better (and more useful).
